# Отчет

## Импорт библиотек и открытие данных

In [1]:
import numpy as np
from prepare_report import knapsack_report, salesman_report
from prepare_report import res_table_salesman, res_table_knapsack
from data.get_branches import get_salesman, get_knapsack

In [ ]:
!unzip /content/genetic

In [2]:
capacity, profits, weights, optimal = get_knapsack('data/benchmarks/')
graph = get_salesman('data/output_graphs.npy')

## Результаты knapsack 0-1

In [ ]:
total_time = []
for i in range(10):
    time_knapsack, weight_knapsack,profit_knapsack, id_knapsack = knapsack_report(capacity, weights, profits)
    total_time.append(time_knapsack)

In [ ]:
total_time = np.array(total_time)
total_time = np.mean(total_time, axis=0)

In [ ]:
res_table_knapsack(sorted(capacity.keys()), total_time, 
                   weight_knapsack, profit_knapsack, 
                   id_knapsack)

p01              p02                 p03  \
Weight                             165               26                 190   
Profit                             309               51                 150   
Time                          0.079288         0.061499            0.061475   
ID      [1, 1, 1, 1, 0, 1, 0, 0, 0, 0]  [0, 1, 1, 1, 0]  [1, 1, 0, 0, 1, 0]   

                          p04                       p05  \
Weight                     50                       104   
Profit                    107                       900   
Time                 0.067658                  0.064515   
ID      [1, 0, 0, 1, 0, 0, 0]  [1, 0, 1, 1, 1, 0, 1, 1]   

                          p06                                            p07  
Weight                    169                                            749  
Profit                   1735                                           1458  
Time                 0.067684                                       0.080929  
ID      [0, 1, 0, 1, 0, 0, 1]  [1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1]

**Предыдущая таблица (лаб2)**

| Рюкзаки\Ответ 	| Вес 	| Стоимость 	| Время 	| ID 	|
|---|---|---|---|---|
| p01 	| 165   	| 309   	| 0.001335 	| [1, 1, 1, 1, 0, 1, 0, 0, 0, 0] 	|
| p02 	| 26    	| 51    	| 0.000094 	| [0, 1, 1, 1, 0] 	|
| p03 	| 190   	| 150    	| 0.001033  	| [1, 1, 0, 0, 1, 0] 	|
| p04 	| 50    	| 107    	| 0.000238  	| [1, 0, 0, 1, 0, 0] 	|
| p05 	| 104   	| 900   	| 0.000658  	| [1, 0, 1, 1, 1, 0, 1, 1] 	|
| p06 	| 169   	| 1735   	| 0.000863 	| [0, 1, 0, 1, 0, 0, 1] 	|
| p07 	| 749   	| 1458   	| 0.010723 	| [1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1] 	|

## Результаты коммивояжер

In [3]:
sorted_keys = ['gr17', 'bays29', 'att48', 'ch150', 'a280', 'fl417']
# paramet = {'gr17': (100, 3_000_000, 10, 0.2), 'bays29': (100, 3_000_000, 10, 0.2), 
#           'att48': (100, 2_000_000, 10, 0.2), 'ch150': (100, 1_000_000, 10, 0.2),
#           'a280': (100, 700_000, 10, 0.2), 'fl417': (100, 500_000, 10, 0.2)}
paramet = {'gr17': (100, 10_000, 20), 'bays29': (100, 100_000, 20), 
          'att48': (100, 100_000, 20), 'ch150': (100, 100_000, 20),
          'a280': (100, 100_000, 20), 'fl417': (100, 100_000, 20)}

In [4]:
time_salesman = []
weight_salesman = []
way_salesman = []
answer_salesman = []

In [5]:
for i in sorted_keys:
    result = salesman_report(graph[i], paramet[i]) 
    
    time_salesman.append(result[3])
    weight_salesman.append(result[1])
    way_salesman.append(result[0])
    answer_salesman.append(result[2])
    
    print(f"{i}: {result[1]} vs {result[2]}")

100%|████████████████████████████████████| 10000/10000 [00:20<00:00, 488.74it/s]


gr17: 2085.0 vs 2085.0


100%|██████████████████████████████████| 100000/100000 [04:23<00:00, 379.30it/s]


bays29: 2092.0 vs 2020.0


100%|██████████████████████████████████| 100000/100000 [06:16<00:00, 265.26it/s]


att48: 12474.0 vs 10628.0


100%|██████████████████████████████████| 100000/100000 [14:28<00:00, 115.11it/s]


ch150: 10136.481391191483 vs 6528.0


100%|███████████████████████████████████| 100000/100000 [24:19<00:00, 68.52it/s]


a280: 2799.3609828948975 vs 2579.0


100%|███████████████████████████████████| 100000/100000 [35:46<00:00, 46.59it/s]


fl417: 29613.841302871704 vs 11861.0


In [6]:
res_table_salesman(sorted_keys, time_salesman, 
          weight_salesman, answer_salesman, way_salesman)

gr17   
Weight                                                  2085.0  \
True answer                                             2085.0   
Time                                                 20.790627   
Way          [2, 10, 9, 1, 4, 8, 11, 15, 0, 3, 12, 6, 7, 5,...   

                                                        bays29   
Weight                                                  2092.0  \
True answer                                             2020.0   
Time                                                264.532747   
Way          [7, 26, 22, 6, 24, 12, 20, 0, 27, 5, 11, 8, 4,...   

                                                         att48   
Weight                                                 12474.0  \
True answer                                            10628.0   
Time                                                 380.00275   
Way          [25, 3, 34, 44, 9, 23, 47, 4, 22, 42, 16, 26, ...   

                                                         ch150   
Weight                                            10136.481391  \
True answer                                             6528.0   
Time                                                906.350482   
Way          [47, 72, 75, 33, 86, 0, 81, 94, 106, 102, 97, ...   

                                                          a280   
Weight                                             2799.360983  \
True answer                                             2579.0   
Time                                               1497.321867   
Way          [1, 279, 0, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 1...   

                                                         fl417  
Weight                                            29613.841303  
True answer                                            11861.0  
Time                                               2181.376102  
Way          [7, 0, 9, 15, 20, 19, 12, 1, 3, 2, 8, 10, 14, ...

In [7]:
for i in way_salesman:
    print(i)

[2, 10, 9, 1, 4, 8, 11, 15, 0, 3, 12, 6, 7, 5, 16, 13, 14]
[7, 26, 22, 6, 24, 12, 20, 0, 27, 5, 11, 8, 4, 25, 28, 2, 1, 19, 9, 3, 14, 17, 16, 13, 21, 10, 18, 15, 23]
[25, 3, 34, 44, 9, 23, 47, 4, 22, 42, 16, 26, 18, 36, 5, 29, 27, 35, 6, 17, 43, 30, 37, 7, 21, 2, 13, 12, 24, 38, 31, 20, 46, 19, 32, 45, 14, 11, 10, 39, 8, 0, 15, 40, 33, 28, 1, 41]
[47, 72, 75, 33, 86, 0, 81, 94, 106, 102, 97, 6, 83, 7, 123, 96, 4, 99, 142, 84, 131, 136, 101, 113, 98, 18, 36, 1, 107, 69, 134, 49, 85, 28, 80, 57, 54, 64, 141, 17, 74, 25, 145, 55, 82, 140, 24, 119, 41, 46, 109, 19, 50, 42, 59, 65, 16, 139, 116, 128, 144, 63, 43, 111, 135, 12, 105, 90, 118, 67, 127, 44, 70, 58, 91, 53, 137, 45, 89, 92, 34, 88, 29, 95, 51, 110, 15, 132, 76, 121, 13, 79, 71, 73, 122, 30, 26, 143, 146, 48, 100, 126, 68, 35, 117, 23, 11, 38, 56, 40, 37, 22, 31, 66, 115, 129, 147, 10, 60, 39, 52, 108, 133, 32, 125, 104, 130, 138, 27, 5, 8, 14, 77, 114, 3, 103, 149, 20, 78, 87, 93, 120, 9, 112, 61, 148, 124, 21, 2, 62]
[1, 279, 0